src/main/java/org/opensearch/ad/task/ADBatchTaskRunner.java

```diff
@@ -244,11 +601,13 @@ private void dispatchTask(ADTask adTask, ActionListener<DiscoveryNode> listener)
                 .collect(Collectors.toList());
 
             if (candidateNodeResponse.size() == 0) {
-                String errorMessage = "All nodes' memory usage exceeds limitation"
-                    + DEFAULT_JVM_HEAP_USAGE_THRESHOLD
-                    + ". No eligible node to run detector "
-                    + adTask.getDetectorId();
-                logger.warn(errorMessage);
+                StringBuilder errorMessageBuilder = new StringBuilder("All nodes' memory usage exceeds limitation ")
```

Comment:
Thanks Kaituo for the suggestion. Explain more about the "one HCAD historical detector at one time problem" with an example, so others can understand what we are talking about. For example, user has a single data nodes cluster, max running tasks per data node is 2 and max running entities per HC detector is also 2. Then if user starts one HC detector, it will use up all 2 task slots on the single data node. So user can't start historical analysis for another detector as there is no available task slot on data node.


Here is some my thoughts
1. Like you said in the cons, this proposal looks like some work around to solve one problem but introduce another problem. If choose this proposal, user needs to increase max task slots per data node to speed up the HCAD historical analysis running. If we keep current design in this PR, user need to reduce the max running entities per detector to support multiple HC historical analysis running. I think they are 50/50 unless we know majority users like running multiple HC historical analysis more than running HC historical analysis faster. When I prototyped HC historical analysis, people(SDM, PM) told me I should prioritize the speed which user cares more, it's not acceptable to let user wait for hours.
2. This proposal is to split entities evenly to different data nodes? One challenging problem in historical analysis(actually in AD and every other plugin) is how to make sure the load is balanced on all data nodes. If we determine which group of entities go to which data node, we can make sure the load is balanced when we split the groups, but we can't guarantee the load will be balanced later. For example, some entities may have much more data, if we put such "big" entities to same data node, that may bring huge load to the data node. 
This proposal needs to sync entity task state regularly, otherwise user won't know the progress. This may bring more network rounds(get task state on each data node) or index querying(index query needs network rounds and disk IO) if the entity task runs for a long time. 
This proposal will add more code/logic branches like queue, regular sync up and error handling for these. Don't think it will help reduce the ops load and code maintenance.

Reply:
Agree with 1.

About 2:
Yes, split entities evenly (statistically) to different data nodes.
Agree, split using hash is worse than your split based on real-time load.  Comparing the worst case of a bad entity vs support one HCAD historical run at one time, the latter is more severe and common.
Is one sync per minute bad enough to cause performance issue?  Also, how do you keep current detector status current?  Do you need to read state index from time to time?  If yes, how is my proposal different?
Could you be more specific that what more branches are needed?  Want to quantify that.

Reply:
For proposal 2
> Comparing the worst case of a bad entity vs support one HCAD historical run at one time, the latter is more severe and common

I feel this is question about how we define&prioritize user experience. Will confirm with PM and SDM. If the first priority is to support multiple HCAD historical run and current solution to let user tune max tasks per node and max entities per detector is not acceptable, we can discuss more tech details about this proposal.

> Is one sync per minute bad enough to cause performance issue? Also, how do you keep detector status current?
I don't know if the "one sync per minute" is correct frequency or not, I think user needs to know the latest status in short time, maybe 5/10seconds, not UX expert, may need to confirm with UX designer. 
Current design is sending back entity task done transport message to update HC detector level task progress and error message. So if some task done (FAILED/FINISHED/STOPPED), we will send message and update HC detector level task immediately. The delay should be less than 1 second.

> Do you need to read state index from time to time? If yes, how is my proposal different?
Yes, the frontend will get HC detector level task doc by task id to get current progress, status etc. In proposal 2, frontend needs to scan all entity tasks to count how many entity tasks done for one specific HC detector, then calculate its task progress and status. 
So the difference is current design just get one doc with doc id from index, proposal 2 need to scan all entity tasks.

> Could you be more specific that what more branches are needed? Want to quantify that.
In proposal 2, the coordinating node just split all entities into groups and send to different data nodes. Then each data node will run entities by itself without talking with other data/master nodes. When each task finished, will update entity task doc. If need to get detector level task state/progress, need to read/scan the task index to get latest data.
In current design, the coordinating node will maintain the entity task queue and dispatch each entity to least-load data node(worker node) with latest entity task state. The data/worker node only maintain its entity task run and send back entity task done message to coordinating node. So coordinating node can gather all entity task state and update HC detector level task progress/status etc. So coordinating node actions like master node of OpenSearch, it actions as some central role and knows everything about entity tasks. 

In proposal 2, we have no some central node, need to read task index to know all entity tasks' state on whole cluster. But the index doc is not always the latest version. It's ok to read this to build some UI feature, as the UI can tolerant  second level latency. But will be risky if we depends on index doc to control resource and debug issues. For example, some node may run 10 entities currently, but just haven't refreshed the task state as `RUNNING` yet. Then if we read index, we may get 5 `RUNNING` entities. If the max running entities limit is 10, then we can still run 5 more entities. That may break our resource control design and we can't trust the running entity number if we use index data in profile API. As the unrefreshed doc is deterministic, it's hard to debug how many entities are running and why there are more entities running for one HC detector. 

And proposal 2 may have horizontal scalability issue. If user adds more data nodes into cluster, then more data nodes will run entity task concurrent queue, and each data node doesn't know how many other entity tasks running on other data nodes. So there may be much more entity tasks running in the cluster, which may bring high load.

Reply:
I agree some and disagree with other arguments. As long as SDM/PM are on the same page, I am fine with current approach. May need to revisit this when we move to serverless.

Reply:
Thanks Kaituo. I think it's good discussion. Let's have a call tomorrow to check if we can have some follow up action to enhance current approach.

Reply:
sure.  Thanks for the discussion.